File: modules/ai/wenxin.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// Wenxin / Baidu ERNIE provider
// ---------------------------------------------------------------------------

const defaultWenxinBaseURL = "https://aip.baidubce.com/rpc/2.0/ai_custom/v1/wenxinworkshop"

// WenxinProvider implements AiProvider for the Baidu ERNIE (Wenxin) API.
// It uses OAuth token-based authentication.
type WenxinProvider struct {
	config     ProviderConfig
	client     *http.Client
	baseURL    string

	mu          sync.Mutex
	accessToken string
	tokenExpiry time.Time
}

// NewWenxinProvider creates a new Wenxin provider with the given config.
func NewWenxinProvider(config ProviderConfig) *WenxinProvider {
	baseURL := config.BaseURL
	if baseURL == "" {
		baseURL = defaultWenxinBaseURL
	}
	return &WenxinProvider{
		config:  config,
		client:  &http.Client{Timeout: config.Timeout()},
		baseURL: baseURL,
	}
}

// Name returns "wenxin".
func (p *WenxinProvider) Name() string { return "wenxin" }

// ---------------------------------------------------------------------------
// Wenxin response wire types
// ---------------------------------------------------------------------------

type wenxinResponse struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	Result  string `json:"result"`  // Wenxin uses "result" instead of "content"
	Model   string `json:"model"`
	Usage   wenxinUsage `json:"usage"`
}

type wenxinUsage struct {
	PromptTokens     uint32 `json:"prompt_tokens"`
	CompletionTokens uint32 `json:"completion_tokens"`
	TotalTokens      uint32 `json:"total_tokens"`
}

// ---------------------------------------------------------------------------
// ChatCompletion (non-streaming)
// ---------------------------------------------------------------------------

// ChatCompletion sends a non-streaming request to the Wenxin API.
func (p *WenxinProvider) ChatCompletion(ctx context.Context, messages []ChatMessage, tools []ToolDefinition, opts ChatOptions) (*AiResponse, error) {
	model := opts.Model
	if model == "" {
		model = p.config.DefaultModel
	}

	reqBody := openAIRequest{
		Model:       model,
		Messages:    messages,
		Tools:       tools,
		Temperature: opts.Temperature,
		MaxTokens:   opts.MaxTokens,
		TopP:        opts.TopP,
		Stop:        opts.StopSequences,
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return nil, NewAiParseError(fmt.Sprintf("failed to marshal request: %s", err))
	}

	accessToken, err := p.getAccessToken(ctx)
	if err != nil {
		return nil, err
	}

	// Wenxin uses model-specific endpoints, e.g., /completions_pro?access_token=xxx
	url := p.buildURL(accessToken, model)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("failed to create request: %s", err))
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("request failed: %s", err))
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, NewAiApiError(
			fmt.Sprintf("API returned status %d: %s", resp.StatusCode, string(respBody)),
			uint16(resp.StatusCode),
		)
	}

	var result wenxinResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, NewAiParseError(fmt.Sprintf("failed to decode response: %s", err))
	}

	return &AiResponse{
		Content: result.Result,
		Model:   result.Model,
		Usage: TokenUsage{
			PromptTokens:     result.Usage.PromptTokens,
			CompletionTokens: result.Usage.CompletionTokens,
			TotalTokens:      result.Usage.TotalTokens,
		},
		FinishReason: "stop",
	}, nil
}

// ---------------------------------------------------------------------------
// ChatCompletionStream (streaming)
// ---------------------------------------------------------------------------

// ChatCompletionStream sends a streaming request to the Wenxin API.
func (p *WenxinProvider) ChatCompletionStream(ctx context.Context, messages []ChatMessage, tools []ToolDefinition, opts ChatOptions) (<-chan StreamEvent, error) {
	model := opts.Model
	if model == "" {
		model = p.config.DefaultModel
	}

	reqBody := openAIRequest{
		Model:       model,
		Messages:    messages,
		Tools:       tools,
		Stream:      true,
		Temperature: opts.Temperature,
		MaxTokens:   opts.MaxTokens,
		TopP:        opts.TopP,
		Stop:        opts.StopSequences,
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return nil, NewAiParseError(fmt.Sprintf("failed to marshal request: %s", err))
	}

	accessToken, err := p.getAccessToken(ctx)
	if err != nil {
		return nil, err
	}

	url := p.buildURL(accessToken, model)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("failed to create request: %s", err))
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("request failed: %s", err))
	}

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		return nil, NewAiApiError(
			fmt.Sprintf("API returned status %d: %s", resp.StatusCode, string(respBody)),
			uint16(resp.StatusCode),
		)
	}

	ch := make(chan StreamEvent, 64)
	go p.readStream(resp, ch)
	return ch, nil
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

// buildURL constructs the Wenxin API URL with the access token.
// The model name maps to a specific endpoint path.
func (p *WenxinProvider) buildURL(accessToken, model string) string {
	endpoint := p.modelToEndpoint(model)
	return p.baseURL + "/" + endpoint + "?access_token=" + accessToken
}

// modelToEndpoint maps a model name to its Wenxin API endpoint path.
func (p *WenxinProvider) modelToEndpoint(model string) string {
	// Common model-to-endpoint mappings for Baidu ERNIE
	switch strings.ToLower(model) {
	case "ernie-4.0", "ernie-4.0-8k", "completions_pro":
		return "completions_pro"
	case "ernie-3.5", "ernie-3.5-8k", "completions":
		return "completions"
	case "ernie-speed", "ernie-speed-8k":
		return "ernie_speed"
	case "ernie-lite", "ernie-lite-8k":
		return "ernie_lite"
	case "ernie-bot-4":
		return "completions_pro"
	default:
		// Default to completions_pro for unknown models
		return "completions_pro"
	}
}

// getAccessToken obtains an access token using the API key and secret.
// It caches the token and refreshes it when expired.
func (p *WenxinProvider) getAccessToken(ctx context.Context) (string, error) {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.accessToken != "" && time.Now().Before(p.tokenExpiry) {
		return p.accessToken, nil
	}

	tokenURL := fmt.Sprintf(
		"https://aip.baidubce.com/oauth/2.0/token?grant_type=client_credentials&client_id=%s&client_secret=%s",
		p.config.APIKey,
		p.config.APISecret,
	)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenURL, nil)
	if err != nil {
		return "", NewAiNetworkError(fmt.Sprintf("failed to create token request: %s", err))
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return "", NewAiAuthError(fmt.Sprintf("token request failed: %s", err))
	}
	defer resp.Body.Close()

	var tokenResp struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int64  `json:"expires_in"`
		Error       string `json:"error,omitempty"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&tokenResp); err != nil {
		return "", NewAiParseError(fmt.Sprintf("failed to decode token response: %s", err))
	}

	if tokenResp.Error != "" {
		return "", NewAiAuthError(fmt.Sprintf("token error: %s", tokenResp.Error))
	}

	if tokenResp.AccessToken == "" {
		return "", NewAiAuthError("empty access token received")
	}

	p.accessToken = tokenResp.AccessToken
	// Set expiry to 30 days minus 5 minutes buffer
	expirySeconds := tokenResp.ExpiresIn
	if expirySeconds <= 0 {
		expirySeconds = 2592000 // 30 days default
	}
	p.tokenExpiry = time.Now().Add(time.Duration(expirySeconds)*time.Second - 5*time.Minute)

	return p.accessToken, nil
}

func (p *WenxinProvider) readStream(resp *http.Response, ch chan<- StreamEvent) {
	defer close(ch)
	defer resp.Body.Close()

	payloads, _ := ParseSSEStream(resp.Body)

	for _, payload := range payloads {
		// Wenxin streaming returns incremental results
		var chunk struct {
			Result string `json:"result"`
			IsEnd  bool   `json:"is_end"`
		}
		if err := json.Unmarshal([]byte(payload), &chunk); err != nil {
			continue
		}

		if chunk.Result != "" {
			ch <- StreamEvent{
				Type:    "content_delta",
				Content: chunk.Result,
			}
		}

		if chunk.IsEnd {
			ch <- StreamEvent{Type: "done"}
			return
		}
	}

	ch <- StreamEvent{Type: "done"}
}

File: packages/ai/src/wenxin.ts
/**
 * @module wenxin
 *
 * Baidu Wenxin (ERNIE) provider implementation.
 *
 * Adapts the Wenxin API to the AiProvider interface.
 * Uses Baidu's authentication flow (access token via API key/secret).
 */
import type { ProviderConfig } from "./provider.js";
import { providerTimeout } from "./provider.js";
import type { ChatMessage, ToolDefinition, ChatOptions, AiResponse, StreamEvent, AiTokenUsage } from "./types.js";
import { newAiParseError, newAiNetworkError, newAiApiError, newAiAuthError } from "./error.js";
import { parseSSEStream } from "./stream.js";

// ---------------------------------------------------------------------------
// Wenxin / Baidu ERNIE provider
// ---------------------------------------------------------------------------
// Baidu's ERNIE (Wenxin) models expose an OpenAI-incompatible HTTP API.
// Two quirks this provider papers over:
//   1. Authentication is OAuth2 (client_credentials) rather than bearer API
//      keys, so we cache an access_token and refresh it ~5 min before expiry.
//   2. The chat completion field is `result` (not `content`), and the
//      endpoint path is per-model (completions_pro vs completions vs
//      ernie_speed/ernie_lite). modelToEndpoint maps friendly names.

const DEFAULT_WENXIN_BASE_URL = "https://aip.baidubce.com/rpc/2.0/ai_custom/v1/wenxinworkshop";

// ---------------------------------------------------------------------------
// Wenxin wire types
// ---------------------------------------------------------------------------
// Shape of the JSON returned by the Wenxin chat endpoints. Only the fields
// we read are typed; the API may return others which we ignore.

interface WenxinResponse {
  id: string;
  object: string;
  result: string; // Wenxin uses "result" instead of "content"
  model: string;
  usage: WenxinUsage;
}

interface WenxinUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

// ---------------------------------------------------------------------------
// WenxinProvider class
// ---------------------------------------------------------------------------

/**
 * Provider implementing AiProvider for Baidu Wenxin/ERNIE models. Owns the
 * OAuth2 access-token lifecycle (cached in-memory, auto-refreshed). Both the
 * non-streaming and streaming chat methods share the same request-body
 * construction; streaming only adds `stream: true` and parses the SSE deltas.
 */

export class WenxinProvider {
  private config: ProviderConfig;
  private baseURL: string;
  private timeoutMs: number;

  // OAuth token state. `tokenExpiry` is a wall-clock ms timestamp; we refresh
  // proactively 5 minutes before it lapses to avoid mid-request 401s.
  private accessToken = "";
  private tokenExpiry = 0;

  constructor(config: ProviderConfig) {
    this.config = config;
    this.baseURL = config.baseURL || DEFAULT_WENXIN_BASE_URL;
    this.timeoutMs = providerTimeout(config);
  }

  name(): string {
    return "wenxin";
  }

  // -------------------------------------------------------------------------
  // ChatCompletion (non-streaming)
  // -------------------------------------------------------------------------

  /**
   * Non-streaming chat completion. Builds the Wenxin request body (note the
   * `result`-style response), waits for the full payload, and normalizes it
   * into the shared {@link AiResponse}. Aborts on timeout.
   */
  async chatCompletion(
    messages: ChatMessage[],
    tools: ToolDefinition[],
    opts: ChatOptions,
  ): Promise<AiResponse> {
    const model = opts.model || this.config.defaultModel;

    const reqBody = {
      model,
      messages,
      tools: tools.length > 0 ? tools : undefined,
      temperature: opts.temperature,
      max_tokens: opts.max_tokens,
      top_p: opts.top_p,
      stop: opts.stop_sequences,
    };

    const body = safeMarshal(reqBody);

    const accessToken = await this.getAccessToken();

    const url = this.buildURL(accessToken, model);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body,
        signal: controller.signal,
      });

      if (!resp.ok) {
        const respBody = await resp.text();
        throw newAiApiError(
          `API returned status ${resp.status}: ${respBody}`,
          resp.status,
        );
      }

      const result = (await resp.json()) as WenxinResponse;

      return {
        content: result.result ?? "",
        model: result.model ?? "",
        usage: {
          prompt_tokens: result.usage?.prompt_tokens ?? 0,
          completion_tokens: result.usage?.completion_tokens ?? 0,
          total_tokens: result.usage?.total_tokens ?? 0,
        },
        finish_reason: "stop",
        tool_calls: [],
      };
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw newAiNetworkError(`request timed out after ${this.timeoutMs}ms`);
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }

  // -------------------------------------------------------------------------
  // ChatCompletionStream (streaming)
  // -------------------------------------------------------------------------

  /**
   * Streaming chat completion. Same request body as {@link chatCompletion}
   * plus `stream: true`; the response body is parsed as SSE where each chunk
   * carries an incremental `result` delta until `is_end` is true.
   */
  async chatCompletionStream(
    messages: ChatMessage[],
    tools: ToolDefinition[],
    opts: ChatOptions,
  ): Promise<AsyncIterable<StreamEvent>> {
    const model = opts.model || this.config.defaultModel;

    const reqBody = {
      model,
      messages,
      tools: tools.length > 0 ? tools : undefined,
      stream: true,
      temperature: opts.temperature,
      max_tokens: opts.max_tokens,
      top_p: opts.top_p,
      stop: opts.stop_sequences,
    };

    const body = safeMarshal(reqBody);

    const accessToken = await this.getAccessToken();

    const url = this.buildURL(accessToken, model);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);

    let resp: Response;
    try {
      resp = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body,
        signal: controller.signal,
      });
    } catch (err) {
      clearTimeout(timer);
      if (err instanceof Error && err.name === "AbortError") {
        throw newAiNetworkError(`request timed out after ${this.timeoutMs}ms`);
      }
      throw err;
    }

    if (!resp.ok) {
      clearTimeout(timer);
      const respBody = await resp.text();
      throw newAiApiError(
        `API returned status ${resp.status}: ${respBody}`,
        resp.status,
      );
    }

    return this.readStream(resp, timer);
  }

  // -------------------------------------------------------------------------
  // Internal helpers
  // -------------------------------------------------------------------------

  /** Constructs the Wenxin API URL with the access token. */
  private buildURL(accessToken: string, model: string): string {
    const endpoint = this.modelToEndpoint(model);
    return `${this.baseURL}/${endpoint}?access_token=${accessToken}`;
  }

  /** Maps a model name to its Wenxin API endpoint path. */
  private modelToEndpoint(model: string): string {
    switch (model.toLowerCase()) {
      case "ernie-4.0":
      case "ernie-4.0-8k":
      case "completions_pro":
        return "completions_pro";
      case "ernie-3.5":
      case "ernie-3.5-8k":
      case "completions":
        return "completions";
      case "ernie-speed":
      case "ernie-speed-8k":
        return "ernie_speed";
      case "ernie-lite":
      case "ernie-lite-8k":
        return "ernie_lite";
      case "ernie-bot-4":
        return "completions_pro";
      default:
        return "completions_pro";
    }
  }

  /**
   * Obtain (or return cached) OAuth2 access token. Uses client_credentials
   * grant with the API key/secret. Caches the token and schedules a refresh
   * 5 minutes before the server-reported expiry to avoid edge-case 401s.
   */
  private async getAccessToken(): Promise<string> {
    if (this.accessToken && Date.now() < this.tokenExpiry) {
      return this.accessToken;
    }

    const tokenURL = `https://aip.baidubce.com/oauth/2.0/token?grant_type=client_credentials&client_id=${this.config.apiKey}&client_secret=${this.config.apiSecret}`;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await fetch(tokenURL, {
        method: "POST",
        signal: controller.signal,
      });

      if (!resp.ok) {
        throw newAiAuthError(`token request failed with status ${resp.status}`);
      }

      const tokenResp = (await resp.json()) as {
        access_token?: string;
        expires_in?: number;
        error?: string;
      };

      if (tokenResp.error) {
        throw newAiAuthError(`token error: ${tokenResp.error}`);
      }

      if (!tokenResp.access_token) {
        throw newAiAuthError("empty access token received");
      }

      this.accessToken = tokenResp.access_token;
      const expirySeconds = tokenResp.expires_in && tokenResp.expires_in > 0
        ? tokenResp.expires_in
        : 2592000; // 30 days default
      this.tokenExpiry = Date.now() + expirySeconds * 1000 - 5 * 60 * 1000; // 5 min buffer

      return this.accessToken;
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw newAiAuthError("token request timed out");
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }

  /**
   * Parse the SSE response body into StreamEvents. Wenxin deltas carry the
   * incremental text in `result`; the stream terminates when a chunk sets
   * `is_end`. The timeout timer is cleared in `finally` so it does not leak
   * if the consumer aborts iteration early.
   */
  private async *readStream(resp: Response, timer: ReturnType<typeof setTimeout>): AsyncGenerator<StreamEvent> {
    try {
      if (!resp.body) {
        yield { type: "done", content: "", tool_call_id: "", tool_call_name: "", arguments: "", usage: null };
        return;
      }

      const reader = resp.body.getReader();
      const payloads = await parseSSEStream(reader);

      for (const payload of payloads) {
        // Wenxin streaming returns incremental results
        let chunk: { result?: string; is_end?: boolean };
        try {
          chunk = JSON.parse(payload) as { result?: string; is_end?: boolean };
        } catch {
          continue;
        }

        if (chunk.result) {
          yield {
            type: "content_delta",
            content: chunk.result,
            tool_call_id: "",
            tool_call_name: "",
            arguments: "",
            usage: null,
          };
        }

        if (chunk.is_end) {
          yield { type: "done", content: "", tool_call_id: "", tool_call_name: "", arguments: "", usage: null };
          return;
        }
      }

      yield { type: "done", content: "", tool_call_id: "", tool_call_name: "", arguments: "", usage: null };
    } finally {
      clearTimeout(timer);
    }
  }
}

/** JSON.stringify wrapper that converts cyclic/unserializable inputs into an AiParseError. */
function safeMarshal(obj: unknown): string {
  try {
    return JSON.stringify(obj);
  } catch (err) {
    throw newAiParseError(`failed to marshal request: ${err instanceof Error ? err.message : String(err)}`);
  }
}

/** Creates a new Wenxin provider with the given config. */
/**
 * Creates a Baidu Wenxin (ERNIE) AI provider.
 *
 * Wenxin uses a token-based authentication flow:
 * 1. Exchange API key/secret for an access token
 * 2. Use the access token for subsequent API calls
 * 3. Handle token refresh when expired
 *
 * @param config - Provider configuration with API key and secret
 * @returns Configured AiProvider for Wenxin/ERNIE models
 */
export function newWenxinProvider(config: ProviderConfig): WenxinProvider {
  return new WenxinProvider(config);
}
